src/text-input/index.tsx

```diff
@@ -106,46 +100,81 @@ function formatAutocomplete(autocomplete: string | boolean | undefined): string
 	return autocomplete;
 }
 
-function patternDiffer(
-	previousProperty: string | undefined,
-	newProperty: string | RegExp | undefined
-): PropertyChangeRecord {
-	const value = newProperty instanceof RegExp ? newProperty.source : newProperty;
-	return {
-		changed: previousProperty !== value,
-		value
-	};
-}
+const factory = create({
+	theme,
+	icache,
+	validity,
+	focus,
+	diffProperty,
+	invalidator
+}).properties<TextInputProperties>();
 
-@theme(css)
-@diffProperty('pattern', patternDiffer)
-@diffProperty('leading', reference)
-@diffProperty('trailing', reference)
-export class TextInput extends ThemedMixin(FocusMixin(WidgetBase))<TextInputProperties> {
-	private _onInput(event: Event) {
-		event.stopPropagation();
-		this.properties.onValue &&
-			this.properties.onValue((event.target as HTMLInputElement).value);
-	}
+export const TextInput = factory(function TextInput({
+	middleware: { icache, theme, validity, focus, diffProperty, invalidator },
+	properties
+}) {
+	diffProperty('pattern', (previous: TextInputProperties, next: TextInputProperties) => {
+		const value = next.pattern instanceof RegExp ? next.pattern.source : next.pattern;
+		if (value !== previous.pattern) {
+			invalidator();
+		}
+	});
+	diffProperty(
+		'leading',
+		(previous: TextInputProperties, next: TextInputProperties) =>
+			previous.leading !== next.leading && invalidator()
+	);
+	diffProperty(
+		'trailing',
+		(previous: TextInputProperties, next: TextInputProperties) =>
+			previous.leading !== next.leading && invalidator()
+	);
 
-	private _onKeyDown(event: KeyboardEvent) {
-		event.stopPropagation();
-		this.properties.onKeyDown &&
-			this.properties.onKeyDown(event.which, () => {
-				event.preventDefault();
-			});
-	}
+	const themeCss = theme.classes(css);
+	const _uuid = icache.getOrSet('uuid', uuid());
+	const dirty = icache.getOrSet('dirty', false);
 
-	private _onKeyUp(event: KeyboardEvent) {
-		event.stopPropagation();
-		this.properties.onKeyUp &&
-			this.properties.onKeyUp(event.which, () => {
-				event.preventDefault();
-			});
-	}
+	const {
+		aria = {},
+		autocomplete,
+		classes,
+		customValidator,
+		disabled,
+		helperText,
+		label,
+		labelHidden = false,
+		leading,
+		max,
+		maxLength,
+		min,
+		minLength,
+		name,
+		onBlur,
+		onClick,
+		onFocus,
+		onKeyDown,
+		onKeyUp,
+		onOut,
+		onOver,
+		onValidate,
+		onValue,
+		pattern: patternValue,
+		placeholder,
+		readOnly,
+		required,
+		step,
+		theme: themeProp,
+		trailing,
+		type = 'text',
+		value,
+		valid: validValue = { valid: undefined, message: '' },
+		widgetId = _uuid
+	} = properties();
+
+	const pattern = patternValue instanceof RegExp ? patternValue.source : patternValue;
 
-	private _callOnValidate(valid: boolean | undefined, message: string) {
-		let { valid: previousValid } = this.properties;
+	function _callOnValidate(valid: boolean | undefined, message: string) {
+		let previousValid: typeof validValue | undefined = validValue;
```

Comment:
Should we re-read properties from here? Like, `const { valid } = properties();`. I think it's a no-no to use properties outside the render function.